snakemake/rules.py

```diff
@@ -1124,25 +1118,54 @@ def expand_group(self, wildcards):
         else:
             return self.group
 
-    def expand_conda_env(self, wildcards):
-        try:
-            conda_env = (
-                self.conda_env.apply_wildcards(wildcards, self)
-                if self.conda_env
-                else None
-            )
-        except WildcardError as e:
-            raise WildcardError(
-                "Wildcards in conda environment file cannot be "
-                "determined from output files:",
-                str(e),
-                rule=self,
+    def expand_conda_env(self, wildcards, params, input):
+        from snakemake.deployment.conda import (
+            is_conda_env_file,
+            CondaEnvFileSpec,
+            CondaEnvNameSpec,
+        )
+
+        if callable(self._conda_env):
+            conda_env, _ = self.apply_input_function(
+                self._conda_env, wildcards, params, input
             )
 
-        if conda_env is not None:
-            conda_env.check()
+            if is_conda_env_file(conda_env):
+                self._conda_env = CondaEnvFileSpec(conda_env, rule=self)
+            else:
+                self._conda_env = CondaEnvNameSpec(conda_env)
+        else:
+            if (
+                conda_env is not None
+                and is_conda_env_file(conda_env)
+                and is_local_file(conda_env)
+                and not os.path.isabs(conda_env)
+            ):
+                conda_env = self.basedir.join(conda_env).get_path_or_uri()
 
-        return conda_env
+            if is_conda_env_file(conda_env):
+                self._conda_env = CondaEnvFileSpec(conda_env, rule=self)
+            else:
+                self._conda_env = CondaEnvNameSpec(conda_env)
```

Comment:
Move this outside of the else and do it for both the callable and the str case. And then, assign the specs to conda_env, not self._conda_env for the same reason as above.